cou/utils/juju_utils.py

```diff
@@ -131,6 +133,67 @@ async def wrapper(*args: Any, **kwargs: Any) -> Any:  # pylint: disable=W9011
     return _wrapper
 
 
+@dataclass(frozen=True)
+class COUMachine:
+    """Representation of a juju machine."""
+
+    machine_id: str
+    hostname: str
+    az: Optional[str] = None  # simple deployments may not have azs
+
+    def __repr__(self) -> str:
+        """Representation of the juju Machine.
+
+        :return: Representation of the juju Machine
+        :rtype: str
+        """
+        return f"Machine[{self.machine_id}]"
+
+
+@dataclass(frozen=True)
+class COUUnit:
+    """Representation of a single unit of application."""
+
+    name: str
+    machine: COUMachine
+    workload_version: str
+
+    def __repr__(self) -> str:
+        """Representation of the application unit.
+
+        :return: Representation of the application unit
+        :rtype: str
+        """
+        return f"Unit[{self.name}]"
+
+
+@dataclass(frozen=True)
+class COUApplication:
+    """Representation of an single application."""
+
+    # pylint: disable=too-many-instance-attributes
+
+    name: str
+    can_upgrade_to: str
```

Comment:
Question: why is this attribute needed?

Reply:
It is defined in `OpenStackApplication`, but never used. So it's not needed.

Reply:
Should I remove it?

Reply:
Just checked, it is being used to determine whether we are at the latest revision of the current channel ([line](https://github.com/canonical/charmed-openstack-upgrader/blob/bc6fe5d89f784d0a8e57056fea9e1297bd9fb8d3/cou/apps/base.py#L392)). So we should leave it.